### ISSUE TITLE
`F.interpolate` uses incorrect size when `align_corners=True`

### ISSUE THREAD

**Issue by arquolo** [pytorch/pytorch#76487]
### 🐛 Describe the bug

Currently for given `scale_factor`  `F.interpolate` uses the same formula to compute output size for both modes of `align_corners`.
For mode `align_corners=False` it's fine, but for `align_corners=True` it's incorrect. as it's makes actual scale factor (defined as scale of image features) dependant from input image size.

Current formula has [off-by-one error](https://en.wikipedia.org/wiki/Off-by-one_error).
For example, for image size of 4 there are 3 (4-1) sections in between pixels. With `scale_factor=2` each section will be populated with 1 new pixel. So correct new size will be 7.
Or, in general, `new_size = (size - 1) * scale_factor + 1`.

Given [this](http://entropymine.com/imageworsener/matching/) `align_corners=False` matches the "match edges" model, and `align_corners=True` - the "match centers" model, but currently scale factor controls scale of image dimensions, not of image features.

This is current behavior:
```
>>> x = torch.rand(1, 1, 4, 4)
>>> y = F.interpolate(x, scale_factor=2, mode='bilinear', align_corners=True)
>>> y.shape
(1, 1, 8, 8)  # incorrect
>>> torch.allclose(x, y[:, :, ::2, ::2])
False
```
Example of correct behavior:
```
>>> y = F.interpolate(x, (7, 7), mode='bilinear', align_corners=True)
>>> torch.allclose(x, y[:, :, ::2, ::2])
True
```

### Versions

Any currently available pytorch version.

**Comment by ngimel** [comment]
cc @vfdev-5, @jbschlosser

**Comment by vfdev-5** [comment]
@arquolo thanks for reporting! I wonder if you have any refs on `new_size = (size - 1) * scale_factor + 1` or could cite any other known projects computing output size according to this formula for `align_corners=True` ? Thanks

Currently, pytorch `align_corners=True` does according the docs:

> If set to ``True``, the input and output tensors are aligned by the center points of their corner pixels, preserving the values at the corner pixels.

So, only corner pixels are aligned. Output image size is computed regardless `align_corner` value:
https://github.com/pytorch/pytorch/blob/f51516df1c977921207a01d601b6915d18bb1286/aten/src/ATen/native/UpSample.cpp#L23-L26

**Comment by arquolo** [comment]
Well, it's pretty hard to find any project using this size formula, as the most of the projects I found either don't align corners (`tf.image.resize` for example), or align but use scale factor as factor to scale dimensions, like `scale` times (`cv2.resize`), but use some padding (see [here](https://jricheimer.github.io/tensorflow/2019/02/11/resize-confusion/)).
Also, I've seen that DeepLab models use interpolation with corner alignment, but they use scale factor instead of size getting input-size dependent scale factor as a result. 

Without corner alignment pytorch works this way:
![image](https://user-images.githubusercontent.com/30900491/165812249-d1fbb0af-f8a0-4095-b420-da0884ebfeea.png)
And with - this way:
![image](https://user-images.githubusercontent.com/30900491/165812380-7beb16b0-6649-46a2-ab99-7c6634e8d51d.png)

My proposal is adjust size formula to get this behavior with corner alignment:
![image](https://user-images.githubusercontent.com/30900491/165812726-01dce4b8-79e5-45c8-96ca-6295c4c60b0f.png)
In this scheme there is 1 pixel outer relative to source image.
And my formula strips those ones.

With corner alignment scale of image is computed this way: `scale = (target size - 1) / (source size - 1)`,
as in this mode pixels are treated as points, and scaling is applied to intervals between them, those count is 1 less then pixel count.
And my formula is just `target size` w.r.t. `scale` and `source size`.

Also, for example with `scale_factor=2`, this size formula will lead to the same result as applying
`nn.ConvTranspose(..., kernel_size=3, stride=2, padding=1)` with kernel `[0.5, 1.0, 0.5]`.

**Comment by vfdev-5** [comment]
> Also, I've seen that DeepLab models use interpolation with corner alignment, but they use scale factor instead of size thus they get scale factor dependent from size of input.

you may be also interested in reading this : https://ppwwyyxx.com/blog/2021/Where-are-Pixels/

> With corner alignment scale of image is computed this way: scale = (target size - 1) / (source size - 1),
as in this mode pixels are treated as points, and scaling is applied to intervals between them, those count is 1 less then pixel count.

You can use `recompute_scale_factor=True` flag to compute target size as source size * scale : 

https://github.com/pytorch/pytorch/blob/8bb720304973f9a22562d9b314edfee4558076fc/torch/nn/functional.py#L3847

**Comment by ppwwyyxx** [comment]
The issue is just an ambiguity of `scale_factor`.

@arquolo said:
> For example, for image size of 4 there are 3 (4-1) sections in between pixels. With scale_factor=2 each section will be populated with 1 new pixel. So correct new size will be 7.

-- This is just __one understanding__ of `scale_factor`: it scales the distance between pixels, aka it scales the size of each section.

However, another understanding of `scale_factor` is that it scales the number of pixels.

As https://ppwwyyxx.com/blog/2021/Where-are-Pixels/#2x-Resize-Operation explains, in `align_corners=True` the two scale factor definitions (stride vs. #pixels) are different. Pytorch chooses the latter definition and there is nothing wrong with that. The documentation also suggests this (though it may use some improvements).
> scale_factor ([float](https://docs.python.org/3/library/functions.html#float) or Tuple[float]) – multiplier for spatial size. 

To achieve what @arquolo expects, you can either use `F.interpolate(x, (7, 7) ...)` or use `scale_factor=7/4`.

**Comment by arquolo** [comment]
I wouldn't call `F.interpolate` with specific size, as I'd like to keep it independent from size of input.
Also, torchscript breaks when I compute size of output at a runtime (i.e. during forward pass). For example like this:
```python
def forward(self, x):
  h, w = [
    (s - 1) * self.scale + 1
    for s in x.shape[2:]]
  return F.interpolate(
    x, (h, w), mode='bilinear', align_corners=True)
```

Given the ambiguity of `scale_factor`, maybe it's better to add `stride` parameter to `F.interpolate`, and internally compute output size using one of them?
